### ISSUE TITLE
Typings: unable to define model with generic Model<T & Document>

### ISSUE THREAD

**Issue by ktutnik** [Automattic/mongoose#10046]
**Do you want to request a *feature* or report a *bug*?**
Report a bug

**What is the current behavior?**

I using a generic repository which I'm not inheriting the document from `Document` class but I use `T & Document` instead. The repository is something like this

```typescript
class Repository<T> {
    private readonly Model: Model<T & Document> // error Type 'T & Document<any, {}>' does not satisfy the constraint 'Document<any, {}>'.

    findById(id:string): Promise<T & Document> {
        return Model.findById(id)
    }
    
    /* other methods */
}
```

This issue may only occur on Mongoose v5.12.1, when I reverted back to version 5.12.0 the issue disappear. 

I reported the same issue #9684 it was fixed, maybe it's missing a test to cover (lock) this feature?

**What are the versions of Node.js, Mongoose and MongoDB you are using? Note that "latest" is not a version.**
Mongoose v5.12.1

**Comment by vkarpov15** [comment]
This is more an issue with how TypeScript handles generics than how Mongoose works. In order to make your syntax work, we would have to make the first parameter to `Model` accept any type, not just documents. See the discussion in https://github.com/microsoft/TypeScript/issues/43281

You should pull the `T & Document` out to where you're using `Repository`, rather than using `T & Document` within a generic class:

```typescript
import { Model, Document } from 'mongoose';

class Repository<T extends Document> {
    private readonly Model: Model<T>

    findById(id:string): Promise<T> {
        return Model.findById(id).exec()
    }
    
    /* other methods */
}

class Foo {
  name: string
}

type Test = Repository<Foo & Document>;
```

**Comment by ktutnik** [comment]
@vkarpov15  no `T & Document` and `T extends Document` is a big different thing. `Repository<T>` means you can use POJO with your repository, it has benefits in several ways. 

Using `Repository<T extends Model>` will be a radical breaking change in my library and everyone using it. https://plumierjs.com/mongoose-helper#domain-model-declaration.

At least, if it's possible to apply this change on major version of mongoose.

**Comment by vkarpov15** [comment]
No idea why, but allowing `__v` to be `any` fixes this :man_shrugging: Thanks for your patience and thanks for reporting this issue, the fix will be in v5.12.4.
